test/test_hessian.jl

```diff
@@ -99,20 +101,38 @@ HESSIAN_TRACERS = (
         ]
 
         h = hessian_sparsity(x -> copysign(x[1] * x[2], x[3] * x[4]), rand(4), method)
-        @test h == [
-            0 1 0 0
-            1 0 0 0
-            0 0 0 0
-            0 0 0 0
-        ]
+        if isshared(T)
+            @test h == [
+                0 1 0 0
+                1 0 0 0
+                0 0 0 1
+                0 0 1 0
+            ]
+        else
+            @test h == [
+                0 1 0 0
+                1 0 0 0
+                0 0 0 0
+                0 0 0 0
+            ]
+        end
 
         h = hessian_sparsity(x -> div(x[1] * x[2], x[3] * x[4]), rand(4), method)
-        @test h == [
-            0 0 0 0
-            0 0 0 0
-            0 0 0 0
-            0 0 0 0
-        ]
+        if isshared(T)
+            @test Matrix(h) == [
+                0 1 0 0
+                1 0 0 0
+                0 0 0 1
+                0 0 1 0
+            ]
+        else
+            @test h == [
+                0 0 0 0
+                0 0 0 0
+                0 0 0 0
+                0 0 0 0
+            ]
+        end
```

Comment:
I think we could return a fully sparse pattern here, even with in-place mutation.

Reply:
See here: https://github.com/adrhill/SparseConnectivityTracer.jl/pull/135/files/3a1df201c6864d50b73359473db058df9b8271d2#diff-438bb091bbff5d5dec5b45b3c61577893cbbd4438cc8fd853135986a34bd8443

Reply:
No we cannot. In shared mode, the resulting pattern is the union of all patterns of intermediate quantities. `x[1] * x[2]` and `x[3] * x[4]` are intermediate quantities